extends Control

@onready var internet: HTTPRequest = $HTTPRequest
@onready var transition = get_tree().current_scene.get_node("Control")
@onready var background = get_tree().current_scene.get_node("Background/Sprite2D")

# Ð²Ñ€ÐµÐ¼ÐµÐ½Ð½Ñ‹Ð¹ URL
var API_URL = "http://127.0.0.1/_api/zrapi/api/v1"
const MOD_ITEM_INTERNET = preload("res://mod_item_internet.tscn")
const LOAD_IMG = preload("res://Resources/loading_img.png")
const MOD_IMG_ERROR = preload("res://image_not_found.png")

var mod_idid = 0
var mods_info

# Called when the node enters the scene tree for the first time.
func _ready() -> void:
	if not transition.imfinished.is_connected(Global.got_finishedsign):
		transition.imfinished.connect(Global.got_finishedsign)
	internet.request(API_URL + "/getmods.php")
	# INFO â†“â†“â†“ Ñ‡Ñ‚Ð¾ ÑÑ‚Ð¾ Ð´ÐµÐ»Ð°ÐµÑ‚? Ð—Ð°Ñ‡ÐµÐ¼ Ð¾Ð½Ð¾? ÐœÑ‹ Ð½Ðµ Ð·Ð½Ð°ÐµÐ¼, lol â†“â†“â†“
	# download("https://kteam.veliona.no/_game_services/zr2.0/mods/files/sus.zip", "user://sus.zip")

# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta: float) -> void:
	pass


func _on_http_request_request_completed(result: int, response_code: int, headers: PackedStringArray, body: PackedByteArray) -> void:
	if response_code == 200:
		mods_info = JSON.parse_string(body.get_string_from_utf8())
		$Control/Panel/HSplitContainer/ScrollContainer/VBoxContainer/Label.queue_free()
		
		for mod in mods_info:
			var new_mod_item = MOD_ITEM_INTERNET.instantiate()
			new_mod_item.IMG = LOAD_IMG
			new_mod_item.NAME = mod["name"]
			new_mod_item.DESC = mod["description"]
			new_mod_item.MOD_ID = mod_idid
			mod_idid += 1

		
			load_image(mod["icon"], new_mod_item, "moditem")
				
		
			$Control/Panel/HSplitContainer/ScrollContainer/VBoxContainer.add_child(new_mod_item)
			print(mod)
	elif response_code == 0:
		$Control/Panel/HSplitContainer/ScrollContainer/VBoxContainer/Label.text = tr("$errorinternetmods0")
	else:
		$Control/Panel/HSplitContainer/ScrollContainer/VBoxContainer/Label.text = tr("$errorinternetmods") % str(response_code)
		


func load_image(url: String, target_node: Node, type: String) -> void:
	var http_request = HTTPRequest.new()
	add_child(http_request)
	http_request.request_completed.connect(_on_image_loaded.bind(http_request, target_node, type))

	http_request.request(url)

func _on_image_loaded(result, code, headers, body, request, target_node, type):
	if result == HTTPRequest.RESULT_SUCCESS:
		var image = Image.new()
		var error = image.load_png_from_buffer(body)
		
		if error == OK:
			var texture = ImageTexture.create_from_image(image)
			if type == "moditem":
				target_node.icon.texture = texture
			elif type == "modinfo":
				target_node.texture = texture
			print(image)
			print(target_node)
		else:
			if type == "moditem":
				target_node.icon.texture = MOD_IMG_ERROR
			elif type == "modinfo":
				target_node.texture = MOD_IMG_ERROR
	
	request.queue_free()


func _on_close_requested() -> void:
	queue_free()
	
func download(url : String, target : String):
	var download_file = target
	$HTTPRequest2.request(url)

func load_mod_info(id):

	mod_idid = id
	#$Control/Panel/HSplitContainer/VSplitContainer/RichTextLabel.text = "\n\n\n[center][tornado]%s[/tornado][/center]" % tr("$loading")
	#
	#http_info.request_completed.connect(dl_mod_info_complete.bind(http_info))
	#http_info.request(API_URL)
	
	dl_mod_info_complete(1, 1, 1, 1, 1) #ÐºÐ¾ÑÑ‚Ñ‹Ð»Ð¸ðŸ˜¨

func dl_mod_info_complete(result, code, headers, body, request):
	
	for mod in mods_info.size():
		if mod == mod_idid:
			$Control/Panel/HSplitContainer/VSplitContainer/RichTextLabel.text = mods_info["data"][mod]["full_description"]
			
			$Control/Panel/HSplitContainer/VSplitContainer/Panel/HBoxContainer/VBoxContainer/modname.text = mods_info["data"][mod]["name"]
			$Control/Panel/HSplitContainer/VSplitContainer/Panel/HBoxContainer/VBoxContainer/moddesc.text = mods_info["data"][mod]["description"]
			
			$Control/Panel/HSplitContainer/VSplitContainer/Panel/HBoxContainer/TextureRect.texture = LOAD_IMG
			load_image(mods_info["data"][mod]["icon"], $Control/Panel/HSplitContainer/VSplitContainer/Panel/HBoxContainer/TextureRect, "modinfo")
			$Control/Panel/HSplitContainer/VSplitContainer/Panel/HBoxContainer/VBoxContainer/dl_btn.disabled = false

func _on_rich_text_label_meta_clicked(meta: Variant) -> void:
	OS.shell_open(str(meta))


func _on_dl_btn_pressed() -> void:
	for mod in mods_info.size():
		if mod == mod_idid:
			$mod_dl.request(mods_info[mod]["file"])
			
			print("ÐÐ°Ñ‡Ð°Ð»Ð¾ Ð·Ð°Ð³Ñ€ÑƒÐ·ÐºÐ¸...")
	
	


func _on_mod_dl_request_completed(result: int, response_code: int, headers: PackedStringArray, body: PackedByteArray) -> void:
	var modname = str(randi_range(100, 1000000))
	var dir = DirAccess.open("user://")
	print("ÐŸÑ€Ð¾Ð²ÐµÑ€ÐºÐ° ÑÑƒÑ‰ÐµÑÑ‚Ð²Ð¾Ð²Ð°Ð½Ð¸Ñ Ð¿Ð°Ð¿Ð¾Ðº, ÑÐ¾Ð·Ð´Ð°Ð½Ð¸Ðµ...")
	if !dir.dir_exists("user://mods"):
		dir.make_dir("user://mods")
	if !dir.dir_exists("user://mods/temp"):
		dir.make_dir("user://mods/temp")
	var file = FileAccess.open("user://mods/temp/temp" + modname + ".zip", FileAccess.WRITE)
	var reader = ZIPReader.new()
	
	if response_code != 200:
		pass
	else:
		print("Ð£ÑÐ¿ÐµÑˆÐ½Ð¾ Ð·Ð°Ð³Ñ€ÑƒÐ¶ÐµÐ½Ð¾, Ð¿Ð¾Ð¿Ñ‹Ñ‚ÐºÐ° ÑÐ¾Ñ…Ñ€Ð°Ð½Ð¸Ñ‚ÑŒ...")
		file.store_buffer(body)
		file.close()
		print("Ð¡Ð¾Ñ…Ñ€Ð°Ð½ÐµÐ½Ð¾! ÐŸÐ¾Ð¿Ñ‹Ñ‚ÐºÐ° Ñ€Ð°ÑÐ¿Ð°ÐºÐ¾Ð²Ð°Ñ‚ÑŒ...")
		
		reader.open("user://mods/temp/temp" + modname + ".zip")
		# ÑƒÐºÑ€Ð°Ð´ÐµÐ½Ð¾ Ñ Ð´Ð¾ÐºÑƒÐ¼ÐµÐ½Ñ‚Ð°Ñ†Ð¸Ð¸ Ð¿Ð¾Ñ‚Ð¾Ð¼Ñƒ-Ñ‡Ñ‚Ð¾ Ñ Ð½ÐµÑƒÑ‡ðŸ˜ŽðŸ˜ŽðŸ˜Ž
		var root_dir = DirAccess.open("user://mods/")

		var files = reader.get_files()
		for file_path in files:
			# If the current entry is a directory.
			if file_path.ends_with("/"):
				root_dir.make_dir_recursive(file_path)
				continue
				
			# Write file contents, creating folders automatically when needed.
			# Not all ZIP archives are strictly ordered, so we need to do this in case
			# the file entry comes before the folder entry.
			root_dir.make_dir_recursive(root_dir.get_current_dir().path_join(file_path).get_base_dir())
			var filee = FileAccess.open(root_dir.get_current_dir().path_join(file_path), FileAccess.WRITE)
			var file2 = DirAccess.open("user://mods/temp/")
			var buffer = reader.read_file(file_path)
			filee.store_buffer(buffer)
			filee.close()
			print("ÐžÑ‡Ð¸ÑÑ‚ÐºÐ°...")
	reader.close()
	print(modname)
	ModLoader.load_mods()
	dir.remove("user://mods/temp/temp" + modname + ".zip")


func _on_quitmodsinternet_pressed() -> void:
#	get_tree().change_scene_to_file("res://mods.tscn")
	background.save_id()
	Global.FROM = 2
	transition.up(3)
